src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -502,18 +493,20 @@ void ExpressionAnalyzer::addJoinAction(ExpressionActionsPtr & actions, JoinPtr j
     actions->add(ExpressionAction::ordinaryJoin(syntax->analyzed_join, join));
 }
 
-bool SelectQueryExpressionAnalyzer::appendJoin(ExpressionActionsChain & chain, bool only_types)
+bool SelectQueryExpressionAnalyzer::appendJoinLeftKeys(ExpressionActionsChain & chain, bool only_types)
 {
-    const ASTTablesInSelectQueryElement * ast_join = getSelectQuery()->join();
-    if (!ast_join)
-        return false;
+    ExpressionActionsChain::Step & step = chain.lastStep(sourceColumns());
+
+    getRootActions(analyzedJoin().leftKeysList(), only_types, step.actions);
+    return true;
+}
 
-    JoinPtr table_join = makeTableJoin(*ast_join);
+bool SelectQueryExpressionAnalyzer::appendJoin(ExpressionActionsChain & chain)
```

Comment:
Result is always `true`. This two methods could return `void` as well.
It is also good to check if we have `JOIN` indeed.

Reply:
There're some methods that do the same. For now `ExpressionAnalyzer::append*()` is expected to return bool. The check if outside of this method, It's private and called from one place, just after the check if we have a join. I'd prefer not to change it in this PR. It's TODO for future refactoring.

Reply:
ok